### PR TITLE
[Feature] Support Charades Dataset

### PR DESCRIPTION
**Pull request by congee524**
- [x] Charades dataset
- [x] Sample Charades Frames
- [ ] Train & Test
- [x] Unit Test
- [x] Prepare Charades dataset
- [ ] Add Comments

### CODE REVIEW

File: mmaction/datasets/pipelines/loading.py [hunk 1]
```diff
@@ -455,6 +455,82 @@ def __repr__(self):
...
+
+        start_index = results['start_index']
+        frame_inds = np.concatenate(frame_inds) + start_index
+        results['frame_inds'] = frame_inds.astype(np.int)
+        results['clip_len'] = self.clip_len
+        results['frame_interval'] = self.frame_interval
+        results['num_clips'] = self.num_clips
+        # aggregate labels of the sampled clip
+        label_list = results['labels'][frame_inds[0] - 1:frame_inds[-1]]
+        label = self.aggregate_labels(label_list)
```
Comment (kennymckormick) [inline mmaction/datasets/pipelines/loading.py:515]: Is it compatible with multiple clip sampling?

File: mmaction/datasets/charades_dataset.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+from .base import BaseDataset
+from .registry import DATASETS
+
+
+@DATASETS.register_module()
+class CharadesDataset(BaseDataset):
+
+    def __init__(self,
+                 ann_file,
```
Comment (dreamerlin) [inline mmaction/datasets/charades_dataset.py:15]: you can use  `*args` and `**kwargs` for simplity

File: mmaction/datasets/charades_dataset.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                frame_labels = row[-1].replace('"', '')
+                if frame_labels != '':
+                    labels[video_name].append(
+                        [int(x) for x in frame_labels.split(',')])
+                else:
+                    labels[video_name].append([])
+
+        if return_list:
+            keys = image_paths.keys()
+            image_paths = [image_paths[key] for key in keys]
```
Comment (dreamerlin) [inline mmaction/datasets/charades_dataset.py:66]: [image_paths[key] for key in image_paths]

File: mmaction/datasets/charades_dataset.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            labels = [labels[key] for key in keys]
+            return image_paths, labels
+        return dict(image_paths), dict(labels)
+
+    def load_annotations(self):
+        """Load annotation file to get video information."""
+        video_infos = []
+        image_paths, labels = self.load_image_lists(self.ann_file)
+        video_names = image_paths.keys()
+        for video_name in video_names:
```
Comment (dreamerlin) [inline mmaction/datasets/charades_dataset.py:76]: `for video_name in image_paths`

File: mmaction/datasets/charades_dataset.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            video_info['total_frames'] = len(image_paths[video_name])
+            video_info['label'] = labels[video_name]
+            if self.filename_tmpl is None:
+                video_info['filename_tmpl'] = self.filename_tmpl_prefix.format(
+                    video_name)
+            video_infos.append(video_info)
+
+        return video_infos
+
+    def prepare_train_frames(self, idx):
```
Comment (dreamerlin) [inline mmaction/datasets/charades_dataset.py:91]: You can directly use those of BaseDataset

File: mmaction/datasets/charades_dataset.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            sample_by_class=sample_by_class,
+            power=power)
+
+    @staticmethod
+    def load_image_lists(frame_list_file, return_list=False):
+        image_paths = defaultdict(list)
+        labels = defaultdict(list)
+        with open(frame_list_file, 'r') as f:
+            assert f.readline().startswith('original_video_id')
+            for i, line in enumerate(f):
```
Comment (dreamerlin) [inline mmaction/datasets/charades_dataset.py:50]: `i` is redundant

File: mmaction/datasets/pipelines/loading.py [hunk 7]
```diff
@@ -455,6 +455,82 @@ def __repr__(self):
         return repr_str
 
 
+@PIPELINES.register_module()
+class SampleCharadesFrames(SampleFrames):
+
+    def __init__(self, clip_len, num_classes=157, **kwargs):
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/loading.py:461]: *args

File: mmaction/datasets/pipelines/loading.py [hunk 8]
```diff
@@ -455,6 +455,82 @@ def __repr__(self):
...
+        results['clip_len'] = self.clip_len
+        results['frame_interval'] = self.frame_interval
+        results['num_clips'] = self.num_clips
+        # aggregate labels of the sampled clip
+        label_list = results['labels'][frame_inds[0] - 1:frame_inds[-1]]
+        label = self.aggregate_labels(label_list)
+        onehot = torch.zeros(self.num_classes)
+        onehot[label] = 1.
+        results['label'] = onehot
+        return results
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/loading.py:519]: `__call__` can inherit from `SampleFrames`

File: mmaction/datasets/pipelines/loading.py [hunk 9]
```diff
@@ -455,6 +455,82 @@ def __repr__(self):
...
+                to the next transform in pipeline.
+        """
+        total_frames = results['total_frames']
+
+        clip_offsets = self._sample_clips(total_frames)
+        frame_inds = clip_offsets[:, None] + np.arange(
+            self.clip_len)[None, :] * self.frame_interval
+        frame_inds = np.concatenate(frame_inds)
+
+        if self.temporal_jitter:
```
[the remark this answers is not in the document]
Comment (congee524) [reply mmaction/datasets/pipelines/loading.py:490]: Here we keep the same with `SampleFrames`

File: mmaction/datasets/charades_dataset.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+            results = copy.deepcopy(self.video_infos[idx])
+        results['filename_tmpl'] = self.filename_tmpl
+        results['modality'] = self.modality
+        results['start_index'] = self.start_index
+
+        return self.pipeline(results)
+
+    def prepare_test_frames(self, idx):
+        """Prepare the frames for testing given the index."""
+        if self.sample_by_class:
```
[the remark this answers is not in the document]
Comment (congee524) [reply mmaction/datasets/charades_dataset.py:107]: May need to keep the same with `Basedataset`

File: mmaction/datasets/charades_dataset.py (hunk 5 above)
Comment (congee524) [reply mmaction/datasets/charades_dataset.py:91]: handling methods of `filename_tmpl` are different.

File: mmaction/datasets/pipelines/loading.py (hunk 8 above)
Comment (congee524) [reply mmaction/datasets/pipelines/loading.py:519]: need to handle `label`

Comment (dreamerlin) [reply mmaction/datasets/pipelines/loading.py:513]:
```python
results = super().__call__(results)
xxxx
```

Comment (congee524) [reply mmaction/datasets/pipelines/loading.py:519]: need `frame_inds`

File: mmaction/datasets/charades_dataset.py (hunk 2 above)
Comment (congee524) [reply mmaction/datasets/charades_dataset.py:16]: Some parameters are not supported by charadesdataset, and we need to modify the default value of multi_class, num_classes. So, maybe better to keep it as it is

File: mmaction/datasets/charades_dataset.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (congee524) [reply mmaction/datasets/charades_dataset.py:107]: `sample_by_class` has been removed in train & test, since our label fromat doesn't support that.

File: mmaction/datasets/charades_dataset.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+    def prepare_train_frames(self, idx):
+        """Prepare the frames for training given the index."""
+        if self.sample_by_class:
+            # Then, the idx is the class index
+            samples = self.video_infos_by_class[idx]
+            results = copy.deepcopy(np.random.choice(samples))
+        else:
+            results = copy.deepcopy(self.video_infos[idx])
+        results['filename_tmpl'] = self.filename_tmpl
```
[the remark this answers is not in the document]
Comment (congee524) [reply mmaction/datasets/charades_dataset.py]: fix

File: mmaction/datasets/charades_dataset.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+@DATASETS.register_module()
+class CharadesDataset(BaseDataset):
+
+    def __init__(self,
+                 ann_file,
+                 pipeline,
+                 data_prefix=None,
+                 test_mode=False,
+                 filename_tmpl_prefix='{}-',
+                 filename_tmpl_suffix='{:06}.jpg',
```
[the remark this answers is not in the document]
Comment (congee524) [reply mmaction/datasets/charades_dataset.py]: cannot format {:06} with '{:06}', maybe the suffix is necessary.

Now, I format {} with '{:06}'

File: mmaction/datasets/pipelines/loading.py (hunk 1 above)
Comment (congee524) [reply mmaction/datasets/pipelines/loading.py]: fixed

File: mmaction/datasets/charades_dataset.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kennymckormick) [reply mmaction/datasets/charades_dataset.py]: That's OK, another possible solution is to use partial: 

from functools import partial
tmpl = '{}_{:05d}.jpg'
tmpl_vid = partial(tmpl.format, 'blahblah')
frame_pth = tmpl_vid(10)

**Review by kennymckormick — COMMENTED** [review]
Would you please update the data preprocessing part, so it would be easier to check the correctness of the codes.

**Comment by congee524** [comment]
> Would you please update the data preprocessing part, so it would be easier to check the correctness of the codes.

The preprocessing script hasn't been implemented yet. I use the slowfast annotations file [here](https://github.com/facebookresearch/SlowFast/blob/master/slowfast/datasets/DATASET.md#charades)

**Comment by congee524** [comment]
kindly ping :p @dreamerlin

File: configs/recognition/slowfast/slowfast_r50_kinetics_8x8x1_50e_charades_rgb.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    test_cfg=dict(maximize_clips='score'))
+
+dataset_type = 'CharadesDataset'
+data_root = 'data/charades/rawframes'
+data_root_val = 'data/charades/rawframes'
+ann_file_train = 'data/charades/annotations/charades_train_list_rawframes.csv'
+ann_file_val = 'data/charades/annotations/charades_val_list_rawframes.csv'
+ann_file_test = 'data/charades/annotations/charades_val_list_rawframes.csv'
+
+mc_cfg = dict(
```
Comment (innerlee) [inline configs/recognition/slowfast/slowfast_r50_kinetics_8x8x1_50e_charades_rgb.py:51]: remove this cfg

File: configs/recognition/slowfast/slowfast_r50_kinetics_cosine_16x8x1_60e_charades_rgb.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    test_cfg=dict(maximize_clips='score'))
+
+dataset_type = 'CharadesDataset'
+data_root = 'data/charades/rawframes'
+data_root_val = 'data/charades/rawframes'
+ann_file_train = 'data/charades/annotations/charades_train_list_rawframes.csv'
+ann_file_val = 'data/charades/annotations/charades_val_list_rawframes.csv'
+ann_file_test = 'data/charades/annotations/charades_val_list_rawframes.csv'
+
+mc_cfg = dict(
```
Comment (innerlee) [inline configs/recognition/slowfast/slowfast_r50_kinetics_cosine_16x8x1_60e_charades_rgb.py:51]: and here

File: mmaction/models/recognizers/base.py [hunk 15]
```diff
@@ -125,40 +125,45 @@ def extract_feat(self, imgs):
             x = self.backbone(imgs)
         return x
 
-    def average_clip(self, cls_score, num_segs=1):
```
Comment (innerlee) [inline mmaction/models/recognizers/base.py:128]: Re-implement `average_clip` using `aggregate_clip` instead of removing it. And add a warning to tell user to use `aggregate_clip`. Also give the link of this pr for reference.

File: mmaction/models/recognizers/base.py [hunk 16]
```diff
@@ -125,40 +125,45 @@ def extract_feat(self, imgs):
...
-            torch.Tensor: Averaged class score.
-        """
-        if 'average_clips' not in self.test_cfg.keys():
-            raise KeyError('"average_clips" must defined in test_cfg\'s keys')
-
-        average_clips = self.test_cfg['average_clips']
+    def aggregate_clip(self, cls_score, num_segs=1):
+        if ('average_clips' not in self.test_cfg) and ('maximize_clips'
+                                                       not in self.test_cfg):
+            raise KeyError(
```
Comment (innerlee) [inline mmaction/models/recognizers/base.py:131]: Does this mean that all configs need to change?

File: mmaction/datasets/pipelines/loading.py [hunk 17]
```diff
@@ -455,6 +455,50 @@ def __repr__(self):
         return repr_str
 
 
+@PIPELINES.register_module()
+class SampleCharadesFrames(SampleFrames):
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:459]: docstring missing.

Pls briefly explain the difference of this class and the existing ones to help users understand why Charades needs a spearate sampling
